numba/inline_closurecall.py

```diff
@@ -236,19 +281,481 @@ def _replace_returns(blocks, target, return_label):
+    try to eliminate array b by expanding a with an extra dimension.
+    """
+    """
+    cfg = compute_cfg_from_blocks(func_ir.blocks)
+    all_loops = list(cfg.loops().values())
+    def find_nest_level(label):
+        level = 0
+        for loop in all_loops:
+            if label in loop.body:
+                level += 1
+    """
+
+    def find_array_def(arr):
+        """Find numpy array definition such as
+            arr = numba.unsafe.ndarray.empty_inferred(...).
+        If it is arr = b[...], find array definition of b recursively.
+        """
+        arr_def = func_ir.get_definition(arr)
+        _make_debug_print("find_array_def")(arr, arr_def)
+        if isinstance(arr_def, ir.Expr):
+            if guard(_find_unsafe_empty_inferred, func_ir, arr_def):
+                return arr_def
+            elif arr_def.op == 'getitem':
+                return find_array_def(arr_def.value)
+        raise GuardException
+
+    def fix_array_assign(stmt):
+        """For assignment like lhs[idx] = rhs, where both a and b are arrays, do the
+        following:
+        1. find the definition of rhs, which has to be a call to numba.unsafe.ndarray.empty_inferred
+        2. find the source array creation for lhs, insert an extra dimension of size of b.
+        3. replace the definition of rhs = numba.unsafe.ndarray.empty_inferred(...) with rhs = lhs[idx]
+        """
+        require(isinstance(stmt, ir.SetItem))
+        require(isinstance(stmt.value, ir.Var))
+        debug_print = _make_debug_print("fix_array_assign")
+        debug_print("found SetItem: ", stmt)
+        lhs = stmt.target
+        # Find the source array creation of lhs
+        lhs_def = find_array_def(lhs)
+        debug_print("found lhs_def: ", lhs_def)
+        rhs_def = _get_definition(func_ir, stmt.value)
+        debug_print("found rhs_def: ", rhs_def)
+        require(isinstance(rhs_def, ir.Expr))
+        if rhs_def.op == 'cast':
+            rhs_def = _get_definition(func_ir, rhs_def.value)
+            require(isinstance(rhs_def, ir.Expr))
+        require(_find_unsafe_empty_inferred(func_ir, rhs_def))
+        # Find the array dimension of rhs
+        dim_def = _get_definition(func_ir, rhs_def.args[0])
+        require(isinstance(dim_def, ir.Expr) and dim_def.op == 'build_tuple')
+        debug_print("dim_def = ", dim_def)
+        extra_dims = [ _get_definition(func_ir, x, lhs_only=True) for x in dim_def.items ]
+        debug_print("extra_dims = ", extra_dims)
+        # Expand size tuple when creating lhs_def with extra_dims
+        size_tuple_def = _get_definition(func_ir, lhs_def.args[0])
+        require(isinstance(size_tuple_def, ir.Expr) and size_tuple_def.op == 'build_tuple')
+        debug_print("size_tuple_def = ", size_tuple_def)
+        size_tuple_def.items += extra_dims
+        # In-place modify rhs_def to be getitem
+        rhs_def.op = 'getitem'
+        rhs_def.value = _get_definition(func_ir, lhs, lhs_only=True)
```

Comment:
The call to `_get_definition` will not raise because a previous `find_array_def(lhs)` (which calls _get_definition, but without setting `lhs_only`) have already returned a value.